src/core/AutomationPattern.cpp

```diff
@@ -460,69 +588,93 @@ void AutomationPattern::flipY()
 
 
 
-void AutomationPattern::flipX( int length )
+void AutomationPattern::flipX(int length)
 {
+	QMutexLocker m(&m_patternMutex);
+
+	timeMap::const_iterator it = m_timeMap.lowerBound(0);
+
+	if (it == m_timeMap.end()) return;
+
+	// Temporary map where we will store the flipped version
+	// of our pattern
 	timeMap tempMap;
 
-	timeMap::ConstIterator iterate = m_timeMap.lowerBound(0);
 	float tempValue = 0;
-	int numPoints = 0;
+	float tempOutValue = 0;
 
-	for( int i = 0; ( iterate + i + 1 ) != m_timeMap.end() && ( iterate + i ) != m_timeMap.end() ; i++)
-	{
-		numPoints++;
-	}
-
-	float realLength = ( iterate + numPoints ).key();
+	// We know the QMap isn't empty, making this safe:
+	float realLength = m_timeMap.lastKey();
 
-	if ( length != -1 && length != realLength)
+	// If we have a positive length, we want to flip the area covered by that
+	// length, even if it goes beyond the pattern. A negative length means that
+	// we just want to flip the nodes we have
+	if (length >= 0 && length != realLength)
 	{
-		if ( realLength < length )
+		// If length to be flipped is bigger than the real length
+		if (realLength < length)
 		{
-			tempValue = valueAt( ( iterate + numPoints ).key() );
-			putValue( MidiTime( length ) , tempValue, false);
-			numPoints++;
-			for( int i = 0; i <= numPoints; i++ )
+			// We are flipping an area that goes beyond the last node. So we add a node to the
+			// beginning of the flipped timeMap representing the value of the end of the area
+			tempValue = valueAt(realLength);
+			tempMap[0] = AutomationNode(this, tempValue, 0);
```

Comment:
I think it would be more confusing. This is called when we are flipping an area that goes beyond the last node of the pattern, so this particular node doesn't exist. We'd need some kind of logic to separate that from the regular node flipping, so I think it isn't worth it the complexity that might be added just for having this inside the loop.

Reply:
Okay! I hadn't really dove into the logic of this, but that makes sense to me. 👍